webapp/graphite/readers/remote.py

```diff
@@ -1,222 +1,94 @@
-                             self.store.host, urlpath)
-        url_full = "%s?%s" % (url, query_string)
         headers = requestContext.get('forwardHeaders') if requestContext else None
 
-        lock = in_flight.get_request_lock(url_full)
-        with lock:
-            request = in_flight.get_request(url_full)
-            if request:
-                log.debug("RemoteReader:: Returning cached FetchInProgress %s" % url_full)
-                return request
-
-            data = self._fetch_list_locked(url, query_string, query_params, headers)
-            in_flight.start_request(url_full, data)
-
-        log.debug(
-            "RemoteReader:: Returning %s in %fs" % (url_full, time.time() - t))
-        return data
-
-    def _fetch_list_locked(self, url, query_string, query_params, headers):
-        url_full = "%s?%s" % (url, query_string)
-
-        jobs = [(self._fetch, url, query_string, query_params, headers)]
-        q = pool_apply(self.store.finder.worker_pool(), jobs)
-
-        log.debug('RemoteReader:: Storing FetchInProgress for %s' % url_full)
-        return FetchInProgress(_Results(q))
-
-    def _fetch(self, url, query_string, query_params, headers):
-        url_full = "%s?%s" % (url, query_string)
-
-        log.debug(
-            "RemoteReader:: Starting to execute _fetch %s" % url_full)
-        try:
-            log.debug("ReadResult:: Requesting %s" % url_full)
-            result = http.request(
-                'POST' if settings.REMOTE_STORE_USE_POST else 'GET',
-                url,
+        retries = 1 # start counting at one to make log output and settings more readable
+        while True:
+          try:
+            result = self.finder.request(
+                '/render/',
                 fields=query_params,
                 headers=headers,
                 timeout=settings.REMOTE_FETCH_TIMEOUT,
             )
-
-            if result.status != 200:
-                self.store.fail()
-                self.log_error("ReadResult:: Error response %d from %s" % (result.status, url_full))
-                data = []
+            break
+          except Exception:
+            if retries >= settings.MAX_FETCH_RETRIES:
+              log.exception("Failed after %s attempts! Root cause:\n%s" %
+                  (settings.MAX_FETCH_RETRIES, format_exc()))
+              raise
             else:
-                data = unpickle.loads(result.data)
-        except Exception as err:
-            self.store.fail()
-            self.log_error("ReadResult:: Error requesting %s: %s" % (url_full, err))
-            data = []
-
-        log.debug("RemoteReader:: Completed _fetch %s" % url_full)
-        return data
-
-
-class _Results(object):
+              log.exception("Got an exception when fetching data! Try: %i of %i. Root cause:\n%s" %
+                           (retries, settings.MAX_FETCH_RETRIES, format_exc()))
+              retries += 1
```

Comment:
Retry logic for remote fetches was moved here, which allows each remote host to be retried individually.